### ISSUE TITLE
Improve devtools / debugging experience / localhost development for Pi Apps running SDK v2

### ISSUE THREAD

**Issue by Sten435** [pi-apps/pi-platform-docs#11]
_How do I debug my application on my MacBook (Chrome) instead of the pi browser._

I want to debug my app using my chrome browser, but I only know a way to open my app using my phone.

How can I open my app in chrome on my pc?

**Comment by dannybutterfield1880** [comment]
The only way I have found so far is using usb debugging on your phone. I use Android ADB to connect to my phone then access chrome://inspect#devices there you can set up port forwarding to your local development environment

**Comment by dannybutterfield1880** [comment]
That is of cause for Android I'm not 100% sure about ios but there should be a similar way of usb debugging and port forwarding

**Comment by aurelienshz** [comment]
Thanks for the issue! We're working on this. There used to be a sandbox environment for the Pi SDK, and we haven't adapted it to work with SDK v2. We're working on providing better developer resources: a demo app with example code, and a decent development environment tooling (either a sandbox or another similar tool).

**Comment by Sten435** [comment]
_I will leave this open for people who have the same issue._

@aurelienshz if you know more about the development, feel free to put it here. 

👍

**Comment by Madhurmila44** [comment]
Thanks developer team for your hard work

**Comment by aurelienshz** [comment]
@Sten435

**Comment by aurelienshz** [comment]
One thing I'd like to add here is that you should be able to load an app served on localhost in the Pi Browser, by using a tunneling service to connect a domain name to your local development server, and registering your tunneled domain name on the developer portal (develop.pi in the Pi Browser). There's a list of tools that enable this over at https://github.com/anderspitman/awesome-tunneling.

I am, however, fully aware that this isn't a perfect experience (e.g you wouldn't be able to open a JS console). And we're working on fixing this!

**Comment by Sten435** [comment]
> The only way I have found so far is using USB debugging on your phone. I use Android ADB to connect to my phone then access chrome://inspect#devices there you can set up port forwarding to your local development environment

I tried this but I can't get it to work in the pi browser. It works on a chrome application on my android phone but when I open the pi browser it does not work.

**Comment by Sten435** [comment]
@dannybutterfield1880 
I tried this but I can't get it to work in the pi browser. It works on a chrome application on my android phone but when I open the pi browser it does not work.

**Comment by dannybutterfield1880** [comment]
I'm actually having an issue using localhost and port forwarding as it means the URL doesn't match in the Dev portal so the request failed anyway and I get a cors error

**Comment by Sten435** [comment]
> @aurelienshz has anyone tried setting URL in the Dev portal to localhost?

Can you change the url in dev portal?

**Comment by dannybutterfield1880** [comment]
@Sten435 yeah, if you go to develop.pi and select your project at the bottom there will be a cog next to the name there it says "the URL that you will host your app"

**Comment by Sten435** [comment]
Yea but, when you change it from a domain to for example localhost, how can the android ADB know that ur using the pi browser and not the chrome one? My dev tools on chrome only detect the activity of URL's that I browse using the chrome app and not using the pi browser.

greetngs

**Comment by dannybutterfield1880** [comment]
@Sten435 it shouldn't matter to ADB, when debugging Pi browser comes right under all of the chrome tabs in its own section. I can't get on my laptop at the moment but I'll send some screen shots over when I can

**Comment by Sten435** [comment]
> @Sten435 it shouldn't matter to ADB, when debugging Pi browser comes right under all of the chrome tabs in its own section. I can't get on my laptop at the moment but I'll send some screen shots over when I can 

Thanks, i will try to do the same
 Greetings.

**Comment by dannybutterfield1880** [comment]
Scratch that. If you try localhost it says invalid URL and if you try something like localhost:8080 it just reloads but doesn't save and goes back to what it was before

**Comment by dannybutterfield1880** [comment]
Maybe look into ngrok it's another tunning service but has a built in UI to see all requests 

https://stackoverflow.com/questions/62644228/how-can-i-view-console-logs-in-ngrok

**Comment by Sten435** [comment]
> Maybe look into ngrok it's another tunning service but has a built in UI to see all requests and also you can see console.log then 
> 
> https://stackoverflow.com/questions/62644228/how-can-i-view-console-logs-in-ngrok

Have you used that with the pi browser before?

**Comment by dannybutterfield1880** [comment]
No I will try shortly though. My girlfriend is using my laptop watching something atm 🙄😂

**Comment by dannybutterfield1880** [comment]
Should work fine, it gives you a URL that you could put in that URL section in develop.pi and acts like a normal web server but tunnels to a port of your choice from your localhost

**Comment by Sten435** [comment]
A i get it, i will also try it. But i never used such a service before...

**Comment by dannybutterfield1880** [comment]
Honestly its easy I believe if you have npm installed you can use npm install -g ngrok

**Comment by dannybutterfield1880** [comment]
Then simply ngrok http PORT

**Comment by Sten435** [comment]
O nice you can use it with npm, i will look it up im verry familliar with that.

**Comment by dannybutterfield1880** [comment]
Nice! You should be fine then

**Comment by Sten435** [comment]
Il keep you updated 👍

**Comment by dannybutterfield1880** [comment]
It works! However still not user friendly I'm afraid, its super slow and the URL will only last 2 hours a time and can only have one connection on the free service

**Comment by Sten435** [comment]
Hmm, maybe its good for this our applications ?

**Comment by dannybutterfield1880** [comment]
Its worth a try but it's not a permanent solution

**Comment by browolf** [comment]
Can someone make a pi emulator that returns dummy results to a normal webbrowser?

**Comment by Sten435** [comment]
@browolf 
I think the dev team is working on somthing like that

**Comment by gradosevic** [comment]
Hi guys! 

Since a lot of devs have issues with debugging their apps in Pi Browser,  my idea for the development is to mock the Pi methods for local testing until we have a sandboxed environment where we can make real service calls on our dev machines. This is just an idea, feel free to check it out here and contribute if you find it useful: 
[https://github.com/gradosevic/pi-mock](https://github.com/gradosevic/pi-mock)

Basically, this way we can work on the implementation of the apps and test it with dummy data. When the app is functional enough we can remove the mock library and test it in Pi browser instead.

@CoreTeam If possible, please make the Pi SDK open for contribution, it would be much easier to make it work for all of us.

**Comment by dannybutterfield1880** [comment]
@gradosevic I will try this shortly! This could be a great solution to get everyone going. 

I would suggest checking your environment for production or development and use pi-mock accordingly

**Comment by Sten435** [comment]
Could you make a detailed guide on how to use it?

**Comment by dannybutterfield1880** [comment]
@gradosevic Thanks for that I have got it working. @Sten435 all you need to do is load Pi like you normally did then copy pi-mock.js into your project and all that does is overwrite ` window.Pi.authenticate() ` and ` window.Pi.createPayment() ` temporarily while pi-mock is imported. I am using react so I am loading the Pi SDK into my index.html file then importing pi-mock.js into my root js file.

**Comment by dannybutterfield1880** [comment]
@Sten435 of course you will not get the Pi wallet UI where you insert your passphrase it will skip that whole part out but the idea is when you remove the mock it should work as expected in the Pi browser

**Comment by dannybutterfield1880** [comment]
@browolf is this the sort of thing you was asking about?

**Comment by browolf** [comment]
@dannybutterfield1880 Yes it's great. I am also collecting a list of these sorts of things here: https://www.reddit.com/r/PiNetwork/wiki/devresources

**Comment by dannybutterfield1880** [comment]
@browolf that's great! Thanks

**Comment by dannybutterfield1880** [comment]
@browolf I'm getting an error when trying to reach that Reddit saying they can't reach their CDN servers

**Comment by gradosevic** [comment]
> @gradosevic I will try this shortly! This could be a great solution to get everyone going.
> 
> I would suggest checking your environment for production or development and use pi-mock accordingly

Thanks for the suggestion, there's now a new configuration property "production_domain" where you can tell what is your production domain and it will turn off in production and use real Pi SDK instead. 

There's a new Pi Mock update where the syntax is slightly simplified, and there are some useful debugging messages. Please refer to the docs for more information. Thanks

**Comment by dannybutterfield1880** [comment]
@gradosevic awesome 👍

**Comment by dannybutterfield1880** [comment]
@gradosevic just to let you know is working perfectly for me. I have included it in my project and at the moment is MIT so check it out. https://github.com/dannybutterfield1880/fire-pi-base

It is boilerplate code for building Pi web apps on Firebase

**Comment by Madhurmila44** [comment]
#43

**Comment by dannybutterfield1880** [comment]
@Madhurmila44 I have commented in your issue

**Comment by Madhurmila44** [comment]
Where can I able to aproove domain

**Comment by Mohamedelsaka1** [comment]
🤙🏻

**Comment by Sten435** [comment]
If anyone has questions about the Pi Network SDK or other general coding questions regarding anything related to the Pi Network, then you ask them in the Pi Network Developer Portal: https://discord.gg/qM3fXPq4fK
It is a faster way to connect with people, instead of contact people through email, or through the Pi chat app, which can be a real pain.

This is not an AD, this is an informative post for people who may need a communication server for a question.

**Comment by Sten435** [comment]
@aurelienshz any updates feom the devs? For how they will adress this issue?

**Comment by cprouveu** [comment]
Hello, When I use the sandbox I have this error. What I'm doing wrong ? 

<img width="1099" alt="Capture d’écran 2021-07-14 à 22 12 42" src="https://user-images.githubusercontent.com/21106757/125687648-e3619c00-4ec1-4aa9-9c68-89aade4434ec.png">

**Comment by Sten435** [comment]
I will close this issue for now, because the sandbox is released and working.

Feel free to re-open this if there are questions...
